runtime/Java/src/org/antlr/v4/runtime/atn/ATNState.java

```diff
@@ -184,24 +181,59 @@ else if ( t.isEpsilon() && e.isEpsilon() ) {
 		}
 		if ( !alreadyPresent ) {
 			transitions.add(index, e);
+			recalculateEpsilonOnlyTransitions();
 		}
 	}
 
 	public Transition transition(int i) { return transitions.get(i); }
 
 	public void setTransition(int i, Transition e) {
-		transitions.set(i, e);
+		transitions.remove(i);
+		recalculateEpsilonOnlyTransitions();
+		if (epsilonOnlyTransitions != null && epsilonOnlyTransitions != e.isEpsilon()) {
+			System.err.format(Locale.getDefault(), "ATN state %d has both epsilon and non-epsilon transitions.\n", stateNumber);
```

Comment:
Why is it useful to print this information? It's a pretty common situation, I'd say, no?

Reply:
Should `set` method be consistent with `add` method that has such check? https://github.com/antlr/antlr4/blob/master/runtime/Java/src/org/antlr/v4/runtime/atn/ATNState.java#L166 I think yes that's why I added the check to `set` method.